// mongodb/lib/core/index.js
'use strict';

let BSON = require('mongodb/lib/core/node_modules/mongodb/lib/core/node_modules/bson');
const require_optional = require('mongodb/lib/core/node_modules/mongodb/lib/core/node_modules/require_optional');
const EJSON = require('mongodb/lib/core/utils').retrieveEJSON();

try {
  // Attempt to grab the native BSON parser
  const BSONNative = require_optional('bson-ext');
  // If we got the native parser, use it instead of the
  // Javascript one
  if (BSONNative) {
    BSON = BSONNative;
  }
} catch (err) {} // eslint-disable-line

module.exports = {
  // Errors
  MongoError: require('mongodb/lib/core/error').MongoError,
  MongoNetworkError: require('mongodb/lib/core/error').MongoNetworkError,
  MongoParseError: require('mongodb/lib/core/error').MongoParseError,
  MongoTimeoutError: require('mongodb/lib/core/error').MongoTimeoutError,
  MongoWriteConcernError: require('mongodb/lib/core/error').MongoWriteConcernError,
  mongoErrorContextSymbol: require('mongodb/lib/core/error').mongoErrorContextSymbol,
  // Core
  Connection: require('mongodb/lib/core/connection/connection'),
  Server: require('mongodb/lib/core/topologies/server'),
  ReplSet: require('mongodb/lib/core/topologies/replset'),
  Mongos: require('mongodb/lib/core/topologies/mongos'),
  Logger: require('mongodb/lib/core/connection/logger'),
  Cursor: require('mongodb/lib/core/cursor').CoreCursor,
  ReadPreference: require('mongodb/lib/core/topologies/read_preference'),
  Sessions: require('mongodb/lib/core/sessions'),
  BSON: BSON,
  EJSON: EJSON,
  Topology: require('mongodb/lib/core/sdam/topology'),
  // Raw operations
  Query: require('mongodb/lib/core/connection/commands').Query,
  // Auth mechanisms
  MongoCredentials: require('mongodb/lib/core/auth/mongo_credentials').MongoCredentials,
  defaultAuthProviders: require('mongodb/lib/core/auth/defaultAuthProviders').defaultAuthProviders,
  MongoCR: require('mongodb/lib/core/auth/mongocr'),
  X509: require('mongodb/lib/core/auth/x509'),
  Plain: require('mongodb/lib/core/auth/plain'),
  GSSAPI: require('mongodb/lib/core/auth/gssapi'),
  ScramSHA1: require('mongodb/lib/core/auth/scram').ScramSHA1,
  ScramSHA256: require('mongodb/lib/core/auth/scram').ScramSHA256,
  // Utilities
  parseConnectionString: require('mongodb/lib/core/uri_parser')
};
